### ISSUE TITLE
[BUG]: idrac_os_deployment : IP Address format is invalid

### ISSUE THREAD

**Issue by kendrew71** [dell/dellemc-openmanage-ansible-modules#435]
### Bug Description

When running this module, it returns "IP Address format is invalid" after about half a minute.

### Component or Module Name

idrac_os_deployment

### Ansible Version

Ansible 2.12.2
dellemc.openmanage v4.4.0

### Python Version

Python 3.8
omsdk 1.2.490

### iDRAC/OME/OME-M version

iDRAC 5.10.10.00

### Operating System

RHEL 8.6

### Playbook Used

```
- hosts: {{ run_target }}
  gather_facts: no
  vars:
    installer_url: "http://{{ builder_host }}/images"
    installer_iso: "installer-test01.iso"
  tasks:
  - name: Boot to Network ISO
    dellemc.openmanage.idrac_os_deployment:
      idrac_ip: "{{ ipmi }}"
      idrac_user: "root"
      idrac_password: "{{ ipmi_pass }}"
      share_name: "{{ installer_url }}"
      iso_image: "{{ installer_iso }}"
    delegate_to: localhost
```

### Logs

```
fatal: [run_target -> localhost]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "expose_duration": 1080,
            "idrac_ip": "aa.bb.5.3",
            "idrac_password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "idrac_port": 443,
            "idrac_user": "root",
            "iso_image": "installer-test01.iso",
            "share_name": "http://aa.bb.5.104/images",
            "share_password": null,
            "share_user": null
        }
    },
    "msg": {
        "DeleteOnCompletion": "false",
        "InstanceID": "DCIM_OSDConcreteJob:1",
        "JobName": "BootToNetworkISO",
        "JobStatus": "Failed",
        "Key": "<null_index>",
        "Message": "IP Address format is invalid.",
        "MessageArguments": null,
        "MessageID": "OSD61",
        "Name": "BootToNetworkISO",
        "Status": "Failed",
        "file": "http://aa.bb.5.104/images/installer-test01.iso",
        "retval": false
    }
}
```

```
racadm>>lclog view -n 3 -q 5720
SeqNumber       = 5720
Message ID      = JCP042
Category        = Config
AgentID         = iDRAC
Severity        = Warning
Timestamp       = 2022-08-13 06:52:33
Message         = Job JID_603915533879 failed because IP Address format is invalid.
RawEventData    = 0xB5,0x00,0x00,0x00,0x08,0x00,0x03,0x09,0x03,0x0A,0x02,0x08,0x00,0x00,0x00,0x00,0x01,0x00,0x00,0x00,0x4C,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x81,0x90,0xF7,0x62,0x00,0x00,0x00,0x00,0x60,0x54,0x00,0x00,0x94,0x00,0x00,0x00,0xB1,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x90,0x00,0x00,0x00,0xAA,0x00,0x00,0x00,0x4A,0x49,0x44,0x5F,0x36,0x30,0x33,0x39,0x31,0x35,0x35,0x33,0x33,0x38,0x37,0x39,0x00,0x00,0x00,0x00,0x14,0xCA,0x04,0x64,0x81,0x90,0xF7,0x62,0x00,0x00,0x00,0x00,0x81,0x90,0xF7,0x62,0x00,0x00,0x00,0x00,0x80,0x00,0x00,0x00,0x03,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x4E,0x65,0x77,0x00,0x4F,0x53,0x44,0x3A,0x20,0x42,0x6F,0x6F,0x74,0x54,0x4F,0x4E,0x65,0x74,0x77,0x6F,0x72,0x6B,0x49,0x53,0x4F,0x00,0x4A,0x43,0x50,0x30,0x30,0x30,0x00,0x4E,0x65,0x77,0x00

Message Arg   1 = JID_603915533879
Message Arg   2 = IP Address format is invalid
FQDD            = N/A
--------------------------------------------------------------------------------
SeqNumber       = 5719
Message ID      = JCP027
Category        = Config
AgentID         = iDRAC
Severity        = Information
Timestamp       = 2022-08-13 06:52:33
Message         = The (installation or configuration) job JID_603915533879 is successfully created on iDRAC.
RawEventData    = 0xB5,0x00,0x00,0x00,0x08,0x00,0x03,0x09,0x03,0x0A,0x02,0x08,0x00,0x00,0x00,0x00,0x01,0x00,0x00,0x00,0x4C,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x60,0x54,0x00,0x00,0x94,0x00,0x00,0x00,0xB1,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x90,0x00,0x00,0x00,0xAA,0x00,0x00,0x00,0x4A,0x49,0x44,0x5F,0x36,0x30,0x33,0x39,0x31,0x35,0x35,0x33,0x33,0x38,0x37,0x39,0x00,0x00,0x00,0x00,0x14,0xCA,0x00,0xFE,0x81,0x90,0xF7,0x62,0x00,0x00,0x00,0x00,0x81,0x90,0xF7,0x62,0x00,0x00,0x00,0x00,0x80,0x00,0x00,0x00,0x03,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x4E,0x65,0x77,0x00,0x4F,0x53,0x44,0x3A,0x20,0x42,0x6F,0x6F,0x74,0x54,0x4F,0x4E,0x65,0x74,0x77,0x6F,0x72,0x6B,0x49,0x53,0x4F,0x00,0x4A,0x43,0x50,0x30,0x30,0x30,0x00,0x4E,0x65,0x77,0x00

Message Arg   1 = JID_603915533879
FQDD            = N/A
--------------------------------------------------------------------------------
SeqNumber       = 5718
Message ID      = USR107
Category        = Audit
AgentID         = WSMAN
Severity        = Information
Timestamp       = 2022-08-13 06:52:33
Message         = The operation BootToNetworkISO of the DCIM_OSDeploymentService was performed by root
Message Arg   1 = BootToNetworkISO
Message Arg   2 = DCIM_OSDeploymentService
Message Arg   3 = root
FQDD            = N/A
--------------------------------------------------------------------------------
```

### Steps to Reproduce

ansible-playbook -i <inv_file> -e "run_target=<target_host>" playbook.yml

### Expected Behavior

The system will boot to the specified ISO at the given URL.

### Actual Behavior

The module returns the "IP address format is invalid" error after 38 seconds.

### Additional Information

I am able to run other dellemc.openmanage modules (i.e., idrac_system_info) without any issues. I am also able to curl the URL in the error output from another Linux host on the subnet.

Not sure if it is trying to ping the IP in the URL first. We disallow ping on our subnets, so that may be a problem.

Also, when I run this task against an iDRAC that does not have a sufficient license for remote OS deployment (iDRAC 8 Basic), I get an error message about licensing, so I don't think that is the issue with this one (iDRAC 9 Express).

**Comment by kendrew71** [comment]
I'm really hoping to find an answer to this (even if it's not a solution). It looks like this issue has been brought up twice before, with no resolution (at least, nothing posted).

It looks like the error is coming from iDRAC, and not the Ansible module directly, but I don't understand why it would be giving this error, since there doesn't seem to be anything incorrect about the IP address. Is the address being packed or otherwise transformed before being sent to the iDRAC?

I'm also trying to understand what capability/feature of the iDRAC that the module is using. It doesn't seem to be using Remote File Share or Virtual Media, because when I run a "remoteimage" command or try to "set VirtualMedia" on the iDRAC itself, it gives a licensing error (expected, since RFS and VM require Enterprise), and not "IP is invalid".

Is the Ansible module doing something similar to how the web UI is able to share a local file ISO to the iDRAC? I'm really curious about that, because racadm doesn't seem to have a comparable functionality (at least that I can find). Any help would be greatly appreciated.

**Comment by anupamaloke** [comment]
@kendrew71, any chance you could raise a Support Request (SR) with Dell Tech Support so that we can track this closely. 

> I'm also trying to understand what capability/feature of the iDRAC that the module is using. It doesn't seem to be using Remote File Share or Virtual Media, because when I run a "remoteimage" command or try to "set VirtualMedia" on the iDRAC itself, it gives a licensing error (expected, since RFS and VM require Enterprise), and not "IP is invalid".

The module uses the WS-Man API provided by iDRAC to Boot to the remote network ISO. 
Other option that you can use is to use a virtual media. Following link has example playbooks on how you can insert a ISO image as a virtual media device and then boot from it: https://github.com/dell/redfish-ansible-module/tree/master/playbooks/virtual_media

**Comment by kendrew71** [comment]
I spent some time yesterday digging into the underlying omsdk python. It looks like this specific problem (invalid IP format) is caused by ShareType processing in "omsdk/sdkfile.py". It loops over the "share_type" enum to determine what type of share is being passed to it by using pre-determined regex patterns.

The trouble (for me) is two-fold. First, apparently http URLs are not supported by the Share class at all, so the derived IP address is literally the string "http". Second, NFS is the first share type processed, but its regex pattern is a superset of the others (FTP, TFTP, SCP), so even those always get processed as NFS type. So they have a similar result, in that their IP address has the type pre-pended (e.g.. scp:aa.bb.cc.dd).

So this currently effectively only works for CIFS and NFS shares. It would be nice if http URLs were supported. It would also be nice if local files were supported (which seemed to be possible as I was looking through the omsdk python).

I've verified that if I move NFS to the end of the enum definition on line 52 of "omsdk/sdkfile.py", it will process the listed types properly with FileOnShare. Now I just need to test to see if it can actually boot the host with an scp share name (boot_to_network_iso), since that is the part that actually talks to the iDRAC, and I haven't yet verified whether it does any additional processing of the share name.

NOTE: Thanks for the info about WS-Man, and the redfish ansible link. I would love to be able to use the linked playbook for VirtualMedia, but as I said we have an Express license, not Enterprise, so that doesn't work. I had already tried those steps manually with "racadm set", but I tried the playbook also just to make sure it wasn't somehow doing it differently, but neither works. Both return the error that VirtualMedia is not licensed.

**Comment by kendrew71** [comment]
I tried letting it run boot_to_network_iso, and as I expected, that does it's own checking of the share path. In this case, it returns immediately with the error "OSD74: Unable to invoke the method because the mandatory input parameter 'ShareType' is invalid". It looks like it uses the same Share class and ShareType enum from "omsdk/sdkfile.py", so it doesn't seem like it should be throwing an error.

I had it print out what it received, and it does see the type as SCP. I changed the share name to use ftp, and got the same error. So it seems like this is effectively disallowing anything but CIFS and NFS, but it seems like an erroneous failure, rather than an intentional type check failure. I'm not able to determine how it's doing the type check though, so I can't really say for sure.

The ansible module does say to use CIFS or NFS for the share, even though it seems like omsdk and the iDRAC itself would support others.
